### DISCUSSION TITLE
Make image transforms an abstract "service"

### DISCUSSION THREAD

**Discussion by khalwat** [craftcms/cms#8471] [Ideas] [answered]
Right now, Image Transforms are tightly coupled to the local implementation.

What I propose instead is that image transforms are written as a generic "service" API, which a defined set of capabilities that it can do when transforming images.

Then the "native" image transform system is written as just one potential implementation of the image transform service.

This simple abstraction layer would allow P&T or third parties to then implement image transforms via other services like https://www.imgix.com/ , https://cloudinary.com/ , https://imageoptim.com , etc.

These services are becoming increasingly popular, and beyond the benefits of supporting them, this abstraction layer would likely result in a better native implementaion.

Having a look at how they offer their APIs would be a good thing to look at when refactoring the Craft APIs for this generic image transform layer.

**Comment by brandonkelly** [comment, marked as answer]
We’ve finally gotten around to this for Craft 4! Details in the PR: #10367

**Reply by ghost** [reply]
Fantastic. Is it possible to provide a toy example of how to implement image transformations other than the native ones?
